proposals/0004-FaultTolerance.md

```diff
@@ -0,0 +1,319 @@
+
+## Detailed design
+
+
+### CDI-based approach 
+Use interceptor and annotation to specify the execution and policy configration.
+An annotation of Asynchronous has to be specified for any asynchronous calls. Otherwise, synchronous execution is assumed. 
+The implementation should provide two interceptors, one for synchronous invocation and the other for asynchronous invocation. 
+```
+@Documented
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ ElementType.METHOD, ElementType.TYPE })
+@Inherited
+public @interface Asynchronous {
+
+}
+```
+#### RetryPolicy: A policy to define the retry criteria
+
+An annotation to specify the max retries, delays, maxDuration, Duration unit, jitter, retryOn, bakeOff, fallback etc.
+```
+@Documented
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ ElementType.METHOD, ElementType.TYPE })
+public @interface Retry {
+   
+   /**
+     *
+     * @return The max number of retries. -1 indicates retry forever.
+     * IllegalArgumentException if maxRetries <-1.
+     */
+    int maxRetries() default 3;
+
+    /**
+     * The delay between retries. Defaults to {@link Duration#NONE}.
+     * @return
+     */
+    int delay() default 0;
+
+    /**
+     *
+     * @return the delay unit
+     */
+
+    TimeUnit delayUnit() default TimeUnit.MILLISECONDS;
+
+    /**
+     * @return the maximum duration to perform retries for.
+     */
+    int maxDuration() default 20;
+
+    /**
+     *
+     * @return the duration unit
+     */
+    TimeUnit durationUnit() default TimeUnit.MILLISECONDS;
+
+    /**
+     *
+     * @return the jitter that randomly vary retry delays by. e.g. a jitter of 200 milliseconds
+     * will randomly add betweem -200 and 200 milliseconds to each retry delay.
+     */
+    int jitter() default 200;
+
+    /**
+     *
+     * @return the jitter delay unit.
+     */
+    TimeUnit jitterDelayUnit() default TimeUnit.MILLISECONDS;
+
+    /**
+     * For each retry delay, a randomly portion of the delay multiplied by the jitterFactor will be added or subtracted to the delay.
+     * e.g. a retry delay of 200 milliseconds and a jitter of 0.25 will result in a random retry delay between 150 and 250 milliseconds.
+     * @return the jitter factor.
+     */
+
+    double jitterFactor() default 0.25;
+
+    /**
+     *
+     * @return Specify the failure to retry on
+     */
+    Class<? extends Throwable>[] retryOn() default { Throwable.class };
+
+    /**
+     *
+     * @return Specify the failure to abort on
+     */
+    Class<? extends Throwable>[] aboartOn() default { Throwable.class };
+
+
+}
+
+
+```
+#### CircuitBreaker: a rule to define when to close the circuit
+
+```
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@Target({ ElementType.TYPE, ElementType.METHOD })
+public @interface CircuitBreaker {
+
+        /**
+     * Define the failure criteria
+     * @return the failure exception
+     */
+    Class<? extends Throwable>[] failOn() default Throwable.class;
```

Comment:
By default our API would consider any exception a failure. This option is more about narrowing the types of exceptions that you're interested in breaking on.

A failure could also be something other than an exception, such as a `null` or negative return value. Overall, we want to be able to characterize what represents a failure for the particular use case.